datadog.go

```diff
@@ -0,0 +1,237 @@
+// Package datadog contains a Datadog exporter.
+//
+// This exporter is currently work in progress
+package datadog
+
+// import "go.opencensus.io/exporter/datadog"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+
+	//"go.opencensus.io/internal"
+	"github.com/DataDog/datadog-go/statsd"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// Exporter exports stats to Datadog
+type Exporter struct {
+	opts      Options
+	collector *collector
+	client    *statsd.Client
+}
+
+// Options contains options for configuring the exporter
+type Options struct {
+	// Namespace to prepend to all metrics
+	Namespace string
+
+	// Host for DogStatsD connection
+	Host string
+
+	// Port for DogStatsD connection
+	Port string
+
+	// OnError is the hook to be called when there is
+	// an error occurred when uploading the stats data.
+	// If no custom hook is set, errors are logged.
+	// Optional.
+	OnError func(err error)
+
+	// Tags are global tags added to each metric
+	Tags []string
+}
+
+var (
+	newExporterOnce      sync.Once
+	errSingletonExporter = errors.New("expecting only one exporter per instance")
+)
+
+// NewExporter returns an exporter that exports stats to Datadog
+func NewExporter(o Options) (*Exporter, error) {
+	var err = errSingletonExporter
+	var exporter *Exporter
+	newExporterOnce.Do(func() {
+		exporter, err = newExporter(o)
+	})
+	return exporter, err
+}
+
+func getEndpoint(o Options) string {
+	var endpoint string
+	if o.Host != "" && o.Port != "" {
+		endpoint = o.Host + ":" + o.Port
+	} else {
+		endpoint = "127.0.0.1:8125"
+	}
+
+	log.Printf("Endpoint set at: %v", endpoint)
+	return endpoint
+}
+
+func newExporter(o Options) (*Exporter, error) {
+
+	endpoint := getEndpoint(o)
+
+	client, err := statsd.New(endpoint)
+	if err != nil {
+		log.Fatal(err)
+	}
+	collector := newCollector(o)
+
+	e := &Exporter{
+		opts:      o,
+		collector: collector,
+		client:    client,
+	}
+	return e, nil
+}
+
+// client implements datadog.Client
+type collector struct {
+	opts Options
+
+	// mu guards all the fields.
+	mu sync.Mutex
+
+	skipErrors bool
+
+	// viewData is accumulated and appended on every Export
+	// invocation from stats.
+	viewData map[string]*view.Data
+
+	viewsMu sync.Mutex
```

Comment:
do we need 2 separated mutex? I think we can lock allthethings when adding a new view